projects/04_cube_xr/main.cpp

```diff
@@ -0,0 +1,373 @@
+            -1.0f, 1.0f,-1.0f,   0.0f, 0.0f, 1.0f,
+            -1.0f,-1.0f,-1.0f,   0.0f, 0.0f, 1.0f,
+
+             1.0f, 1.0f,-1.0f,   1.0f, 1.0f, 0.0f,  // +X side
+             1.0f, 1.0f, 1.0f,   1.0f, 1.0f, 0.0f,
+             1.0f,-1.0f, 1.0f,   1.0f, 1.0f, 0.0f,
+             1.0f,-1.0f, 1.0f,   1.0f, 1.0f, 0.0f,
+             1.0f,-1.0f,-1.0f,   1.0f, 1.0f, 0.0f,
+             1.0f, 1.0f,-1.0f,   1.0f, 1.0f, 0.0f,
+
+            -1.0f,-1.0f,-1.0f,   1.0f, 0.0f, 1.0f,  // -Y side
+             1.0f,-1.0f,-1.0f,   1.0f, 0.0f, 1.0f,
+             1.0f,-1.0f, 1.0f,   1.0f, 0.0f, 1.0f,
+            -1.0f,-1.0f,-1.0f,   1.0f, 0.0f, 1.0f,
+             1.0f,-1.0f, 1.0f,   1.0f, 0.0f, 1.0f,
+            -1.0f,-1.0f, 1.0f,   1.0f, 0.0f, 1.0f,
+
+            -1.0f, 1.0f,-1.0f,   0.0f, 1.0f, 1.0f,  // +Y side
+            -1.0f, 1.0f, 1.0f,   0.0f, 1.0f, 1.0f,
+             1.0f, 1.0f, 1.0f,   0.0f, 1.0f, 1.0f,
+            -1.0f, 1.0f,-1.0f,   0.0f, 1.0f, 1.0f,
+             1.0f, 1.0f, 1.0f,   0.0f, 1.0f, 1.0f,
+             1.0f, 1.0f,-1.0f,   0.0f, 1.0f, 1.0f,
+        };
+        // clang-format on
+        uint32_t dataSize = ppx::SizeInBytesU32(vertexData);
+
+        grfx::BufferCreateInfo bufferCreateInfo       = {};
+        bufferCreateInfo.size                         = dataSize;
+        bufferCreateInfo.usageFlags.bits.vertexBuffer = true;
+        bufferCreateInfo.memoryUsage                  = grfx::MEMORY_USAGE_CPU_TO_GPU;
+
+        PPX_CHECKED_CALL(GetDevice()->CreateBuffer(&bufferCreateInfo, &mVertexBuffer));
+
+        void* pAddr = nullptr;
+        PPX_CHECKED_CALL(mVertexBuffer->MapMemory(0, &pAddr));
+        memcpy(pAddr, vertexData.data(), dataSize);
+        mVertexBuffer->UnmapMemory();
+    }
+
+    // Viewport and scissor rect
+    mViewport    = {0, 0, float(GetWindowWidth()), float(GetWindowHeight()), 0, 1};
+    mScissorRect = {0, 0, GetWindowWidth(), GetWindowHeight()};
+}
+
+void ProjApp::Render()
+{
+    PerFrame& frame = mPerFrame[0];
+
+    uint32_t currentViewIndex = 0;
+#if defined(PPX_BUILD_XR)
+    if (GetSettings()->enableXR) {
+        currentViewIndex = GetXrComponent().GetCurrentViewIndex();
+    }
+#endif
+
+    grfx::SwapchainPtr swapchain = GetSwapchain(currentViewIndex);
+
+    uint32_t imageIndex = UINT32_MAX;
+
+#if defined(PPX_BUILD_XR)
+    if (GetSettings()->enableXR) {
+        // no need to
+        // - signal imageAcquiredSemaphore & imageAcquiredFence
+        // - wait for imageAcquiredFence since xrWaitSwapchainImage is called in AcquireNextImage
+        PPX_CHECKED_CALL(swapchain->AcquireNextImage(UINT64_MAX, nullptr, nullptr, &imageIndex));
+    }
+    else
+#endif
+    {
+        // Wait semaphore is ignored for XR
+        PPX_CHECKED_CALL(swapchain->AcquireNextImage(UINT64_MAX, frame.imageAcquiredSemaphore, frame.imageAcquiredFence, &imageIndex));
+
+        // Wait for and reset image acquired fence
+        PPX_CHECKED_CALL(frame.imageAcquiredFence->WaitAndReset());
+    }
```

Comment:
yeah, I am not happy with this neither, but I feel as the first example, it is better to be explicit about not using semaphore, and we could refine the API later.

Reply:
interesting idea. i think it is better than checking application setting since in the future we might have both swapchains from XR and swapchain from glfw window (this is for renderdoc capture), and having the requirement from the swapchain would be clearer for the user.  
I feel that "requireSignaling" is still a bit confusing, but i cannot think of a better name than this for now.

Reply:
Or maybe something even more generic: `requiresExternalSynchronization` or `isThreadSafe` ?

Reply:
k, i will go with `requiresExternalSynchronization` for now, and we can change it later if there is a better option :)

Reply:
i will change it to `ShouldSkipExternalSynchronization` to inverse the logic so that i dont need to add this check for all existing non-xr code. Please let me know if you have a better name, and i am happy to change :)

Reply:
fixed